### DISCUSSION TITLE
Improving Snippets functionality to allow professional work

### DISCUSSION THREAD

**Discussion by NotANormalNerd** [wagtail/wagtail#8403] [Ideas]
I started working with Wagtail professionally at my current company, roughly half a year ago. I got to know the ropes and got to love wagtail. But something is just off when I come by snippets:

The idea is amazing, have something that you can reuse all over your pages and like "Ads" or Authors. Or have something that does not constitute a full page, but still needs to live somewhere. But the execution is... lacking.

What I am mostly missing is:
* Basic publishing ability (at least Draft & Published)
* Workflows (Approvals, Translation)
* Revision History
Extended missing:
* Metadata (for us e.g. which customer are allowed to see a campaign)
* Organisation (per Site)

I think Snippets are really a useful feature and can extend the usability of Wagtail a lot. Sadly especially for bigger organisations those features are showstoppers.

For us this would mostly mean: Marketing creates "campaigns" and the content managers can pickup those campaigns. Content managers cannot change the campaigns and marketing cannot change the content (pages).

Currently we just whacked the Page Type into shape and needed to write some custom code to pull in a page into another page. Which really does not make that much sense.

Maybe we can work on improving the Snippet functionality to actually make the feature what it deserves to be.

Edit: added Revision History

**Comment by cnk** [comment]
Since snippets are not stand alone pages, I don't see how SEO stuff is applicable. But being able to have per site snippets would be super useful.

**Reply by NotANormalNerd** [reply]
I changed the SEO mention and added something from our side.

**Reply by lb-** [reply]
What do you mean by a site Snippet?

There is SiteSettings - could you link a snippet chooser to that?

Just trying to understand what that would look like and what cannot be built with the existing parts of Wagtail?

**Reply by cnk** [reply]
I have lots of sites in the same Wagtail instance - all of which are quite independent of the others. So our Snippet and ModelAdmin models all have a site_id column. I would love to have a built in way to have people restricted to editing snippets from their own site. Similarly I would like the chooser to offer only snippets for the current site. This might not work for sites that have separate public sites but a shared admin side so we would have to figure out if it is something that could/should go in Wagtail itself. But since this is a wish list, I thought I would mention it.

**Reply by phildexter** [reply]
@cnk that sounds similar (to me) to Collections but for Snippets? Being able to segregate certain objects into folders that are relevant to certain people, and can be restricted. Is that understanding correct?

**Reply by cnk** [reply]
@phildexter exactly. For my use case, I would want sets of snippets that belong to a site. But I suspect for large single sites, editors might find it convenient to group snippets by purpose, etc. and then give a group permission over a group of snippets. One option might be to make them inherit from CollectionMember like images and documents do. But I have not thought deeply about this; there may be better options.

**Reply by phildexter** [reply]
Wondering if collections is the right thing here or not...

Thinking about how collections work vs snippets - collections are used for images and documents, that otherwise are pretty much uncategorised. They are generic images and generic documents. A collection is assigned to an individual image or document instance. Is that how you'd want it to work for snippets in your use-case? Or is it closer to, snippets of this type / categorised under this parent object, belong in this permissions group?

**Comment by phildexter** [comment]
@NotANormalNerd I'm not sure if you saw the latest [What's new in Wagtail webinar](https://youtu.be/owsdUze9m0Q), but there we showed off some new functionality for Snippets, which includes some of your wish list. Specifically Draft / Publish and revision history. We may well also do workflows and previews too.
